yarn-project/simulator/src/private/providers/circuit_recording/simulation_provider_recorder_wrapper.ts

```diff
@@ -0,0 +1,85 @@
+import { createLogger } from '@aztec/foundation/log';
+import type { ForeignCallHandler } from '@aztec/noir-protocol-circuits-types/types';
+import type { FunctionArtifactWithContractName } from '@aztec/stdlib/abi';
+import type { NoirCompiledCircuitWithName } from '@aztec/stdlib/noir';
+
+import type { ACIRCallback, ACIRExecutionResult } from '../../acvm/acvm.js';
+import type { ACVMWitness } from '../../acvm/acvm_types.js';
+import type { SimulationProvider } from '../simulation_provider.js';
+import { CircuitRecorder } from './circuit_recorder.js';
+
+/**
+ * Takes a simulation provider and wraps it in a circuit recorder. See CircuitRecorder for more details on how circuit
+ * recording works.
+ */
+export class SimulationProviderRecorderWrapper implements SimulationProvider {
+  constructor(private simulator: SimulationProvider, private log = createLogger('simulator-recorder')) {}
+
+  async executeProtocolCircuit(
+    input: ACVMWitness,
+    artifact: NoirCompiledCircuitWithName,
+    callback: ForeignCallHandler | undefined,
+  ): Promise<ACVMWitness> {
+    const recordingEnabled = process.env.CIRCUIT_RECORD_DIR !== undefined;
+    let circuitRecorder: CircuitRecorder | undefined;
+    let wrappedCallback: ForeignCallHandler | undefined = undefined;
+    if (recordingEnabled) {
+      // Recording is enabled so we start a new circuit recorder
+      circuitRecorder = await CircuitRecorder.start(
+        process.env.CIRCUIT_RECORD_DIR!,
+        input,
+        Buffer.from(artifact.bytecode, 'base64'),
+        artifact.name,
+        'main',
+      );
+      if (callback !== undefined) {
+        // Callback was provided so we wrap it in a circuit recorder callback wrapper
+        wrappedCallback = circuitRecorder.wrapProtocolCircuitCallback(callback);
+      }
+    }
+
+    try {
+      const result = await this.simulator.executeProtocolCircuit(
+        input,
+        artifact,
+        wrappedCallback ? wrappedCallback : callback,
+      );
+
+      if (recordingEnabled) {
+        // Recording is enabled and witness generation is complete so we finish the circuit recorder
+        await circuitRecorder!.finish();
+      }
+
+      return result;
+    } catch (err) {
+      this.log.error('Circuit execution failed', { error: err });
+      throw err;
```

Comment:
Don't we want to finish the file regardless? It'll still be useful, and it's our main usecase (debugging failing circuts). I'd consider a `finally` block.

Reply:
Addressed in https://github.com/AztecProtocol/aztec-packages/pull/12148/commits/732554c700688a9ec3f3335c5572fdc2d4324152